### PR TITLE
WIP: ConfigObjectUtility::CreateObject(): ignore empty files

### LINKED ISSUE #3937: [dev.icinga.com #11157] Icinga2 API: PUT request fails at 0-byte file
This issue has been migrated from Redmine: https://dev.icinga.com/issues/11157

**Created by elabedzki on 2016-02-15 07:57:56 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _Backlog_
Last Update: _2016-11-09 14:54:28 +00:00 (in Redmine)_

    Icinga Version: r2.4.1-1
    Backport?: Not yet backported
    Include in Changelog: 1

---

Hello,

for automated filling of Icinga2 configuration via the API is swamped us the / var file system. After pruning the File system and the start and stop of Icinga we have the following problem.

In the API request if the service group SAP\_System\_DWV exists we get

"Object does not exist." In response.

 

Request:

Mon February 1 13:45:41 2016 URL: https://153.95.26.168:5665/v1/objects/servicegroups/SAP\_System\_DWV?attrs=name

Mon February 1 13:45:41 2016 HTTP headers:

X-HTTP-Method-Override: GET

 

Answer:

Error: Object does not exist.

 

        (0) libremote.so: void boost :: throw\_exception <boost :: :: exception_detail error_info_injector <std :: invalid_argument>&gt; (boost :: :: exception\_detail error\_info\_injector <std :: invalid_argument> const &) (+ 0xE8) \[0x7fcf3004f7c8\]

        (1) libremote.so: void boost :: :: exception\_detail throw\_exception\_ <std :: invalid_argument> (const std :: invalid\_argument &, const char **, char const**, int) (+ 0x59) \[0x7fcf3004f879\]

        (2) libremote.so: icinga :: :: ConfigObjectTargetProvider GetTargetByName (icinga :: string const &, icinga :: string const &) const (+ 0xAD) \[0x7fcf2fff16cd\]

        (3) libremote.so: icinga :: :: FilterUtility GetFilterTargets (icinga :: Query Description const &, boost :: intrusive\_ptr <icinga :: Dictionary> const &, boost :: intrusive\_ptr <icinga :: ApiUser> const &) (+ 0x1f7) \[ 0x7fcf2fff7667\]

        (4) libremote.so: icinga :: Object Query Handler :: HandleRequest (boost :: intrusive\_ptr <icinga :: ApiUser> const &, icinga :: HTTPRequest &, icinga :: HttpResponse &) (+ 0x573) \[0x7fcf30005253\]

        (5) libremote.so: icinga :: :: HTTPHandler ProcessRequest (boost :: intrusive\_ptr <icinga :: ApiUser> const &, icinga :: HTTPRequest &, icinga :: HttpResponse &) (+ 0x516) \[0x7fcf2ffff276\]

        (6) libremote.so: icinga :: Http Server Connection :: ProcessMessageAsync (icinga :: HTTPRequest &) (+ 0x487) \[0x7fcf30036057\]

        (7) libbase.so: icinga :: :: WorkQueue WorkerThreadProc () (+ 0x259) \[0x7fcf32aeb249\]

        (8) libboost\_thread.so.1.54.0: <unknown function> (+ 0xba4a) \[0x7fcf33477a4a\]

        (9) libpthread.so.0: <unknown function> (+ 0x8182) \[0x7fcf32de4182\]

        (10) libc.so.6: clone (+ 0x6d) \[0x7fcf31c50fbd\]

 

 

If we try the service group using the API to create so suggests that the message

"Error: Configuration file '/var/lib/icinga2/api/packages/\_api/fciu03v168-1453127875-1/conf.d/servicegroups/SAP\_System\_DWV.conf' already exists. "

fail.

 

Request:

Mon February 1 13:45:41 2016 Action: "add" Object name: "SAP\_System\_DWV"

icinga2\_api\_write: Action: "add" Object name: "SAP\_System\_DWV" Object type "Service Group"

Mon February 1 13:45:41 2016 URL: https://153.95.26.168:5665/v1/objects/servicegroups/SAP\_System\_DWV

Mon February 1 13:45:41 2016 HTTP headers:

Accept: application / json

X-HTTP-Method-Override: PUT

 

Mon February 1 13:45:41 2016 JSON request:

{ [Attrs]() { [display\_name]() "SAP System DWV"}}

 

Answer:

Error: Configuration file '/var/lib/icinga2/api/packages/\_api/fciu03v168-1453127875-1/conf.d/servicegroups/SAP\_System\_DWV.conf' already exists.

 

 

In the file system, there are an empty file:

root @ fciu03v168: / var / log / Icinga2 / compat / archives \# ll /var/lib/icinga2/api/packages/\_api/fciu03v168-1453127875-1/conf.d/servicegroups/SAP\_System\_DWV.conf

rw-r - r-- 1 nagios nagios 0 February 1 13:00 /var/lib/icinga2/api/packages/\_api/fciu03v168-1453127875-1/conf.d/servicegroups/SAP\_System\_DWV.conf

 

The statement of GET Request "Object does not exist" stands in contradiction to the response of the PUT Requests "\_System\_DWV.conf already exists"

 

We use the following Icinga2 Version:

 

Icinga2 -V

Icinga2 - The Icinga 2 network monitoring daemon (version: r2.4.1-1)

 

Copyright © 2012-2015 Icinga Development Team (https://www.icinga.org)

License GPLv2 +: GNU GPL version 2 or later &lt;http://gnu.org/licenses/gpl2.html&gt;

This is free software: you are free to change and redistribute it.

There is NO WARRANTY, to the extent permitted by law.

 

Application information:

  Installation root: / usr

  Sysconf directory: / etc

  Run directory: / run

  Local state directory: / var

  Package datadirectory: / usr / share / Icinga2

  State path: /var/lib/icinga2/icinga2.state

  Modified attributes path: /var/lib/icinga2/modified-attributes.conf

  Objects path: /var/cache/icinga2/icinga2.debug

  Vars path: /var/cache/icinga2/icinga2.vars

  PID path: /run/icinga2/icinga2.pid

 

System information:

  Platform: \_NAME = "Ubuntu 14.04.1 LTS

  Platform version: ID = "14:04

  Kernel: Linux

  Kernel version: 3.13.0-24-generic

  Architecture: x86\_64

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #3937

### CODE REVIEW

**Comment by Al2Klimov** [comment]
Rationale: https://github.com/Icinga/icinga2/issues/3937#issuecomment-273097131

File: lib/remote/configobjectutility.cpp [hunk 1]
```diff
@@ -112,9 +118,24 @@ bool ConfigObjectUtility::CreateObject(const Type::Ptr& type, const String& full
...
 	Utility::MkDirP(Utility::DirName(path), 0700);
 
-	if (Utility::PathExists(path)) {
-		errors->Add("Configuration file '" + path + "' already exists.");
-		return false;
+	{
+
+#ifdef _WIN32
+#define stat _stat
+#endif /* _WIN32 */
```
Comment (dnsmichi) [inline lib/remote/configobjectutility.cpp]: That's bad coding practice to rename functions. Please remove that and change the underlaying code.

File: lib/remote/configobjectutility.cpp [hunk 2]
```diff
@@ -112,9 +118,24 @@ bool ConfigObjectUtility::CreateObject(const Type::Ptr& type, const String& full
...
+		if (stat(path.CStr(), &stats)) {
+			if (errno != ENOENT)
+				BOOST_THROW_EXCEPTION(posix_error()
+				    << boost::errinfo_api_function("stat")
+				    << boost::errinfo_errno(errno)
+				    << boost::errinfo_file_name(path));
+		} else if (stats.st_size) {
+			errors->Add("Configuration file '" + path + "' already exists.");
+			return false;
+		}
```
Comment (dnsmichi) [inline lib/remote/configobjectutility.cpp]: That will solve the problem with 0 byte files, but what if the file was scrambled with garbage? The real problem hiding here is that config files created from API requests are written "as is", without any temporary file & rename.

Comment (Al2Klimov) [inline lib/remote/configobjectutility.cpp]: I think I'll wait for #5220.

**Comment by dnsmichi** [comment]
Requires a new patch set. Closing here, thanks for the effort.

**Pull request closed without merging**